akka-projection-grpc/src/main/scala/akka/projection/grpc/internal/ProtobufProtocolConversions.scala

```diff
@@ -52,30 +53,51 @@ private[akka] object ProtobufProtocolConversions {
     offset match {
       case None => NoOffset
       case Some(o) =>
-        val timestamp =
-          o.timestamp.map(_.asJavaInstant).getOrElse(Instant.EPOCH)
-        val seen = o.seen.map {
-          case PersistenceIdSeqNr(pid, seqNr, _) =>
-            pid -> seqNr
-        }.toMap
-        TimestampOffset(timestamp, seen)
+        if (o.offsetsBySlice.nonEmpty) {
+          val offsets = o.offsetsBySlice.flatMap { offsetBySlice =>
+            offsetBySlice.offset.map(protocolOffsetToTimestampOffset).map(offsetBySlice.slice.->)
+          }
+          TimestampOffsetBySlice(offsets.toMap)
+        } else {
+          protocolOffsetToTimestampOffset(o)
+        }
     }
 
+  private def protocolOffsetToTimestampOffset(offset: proto.Offset): TimestampOffset = {
+    val timestamp =
+      offset.timestamp.map(_.asJavaInstant).getOrElse(Instant.EPOCH)
+    val seen = offset.seen.map {
+      case PersistenceIdSeqNr(pid, seqNr, _) =>
+        pid -> seqNr
+    }.toMap
+    TimestampOffset(timestamp, seen)
+  }
+
   def offsetToProtoOffset(offset: Offset): Option[proto.Offset] = {
     offset match {
-      case TimestampOffset(timestamp, _, seen) =>
-        val protoTimestamp = Timestamp(timestamp)
-        val protoSeen = seen.iterator.map {
-          case (pid, seqNr) =>
-            PersistenceIdSeqNr(pid, seqNr)
-        }.toSeq
-        Some(proto.Offset(Some(protoTimestamp), protoSeen))
+      case timestampOffset: TimestampOffset =>
+        Some(timestampOffsetToProtoOffset(timestampOffset))
+      case TimestampOffsetBySlice(offsets) =>
+        val offsetsBySlice = offsets.toSeq.map {
```

Comment:
`iterator` instead of `toSeq`?

Reply:
Sure. Can update to `iterator.map.{ ... }.toSeq`. Or maybe iterating on the iterable once directly (just `offsets.map`) is better?